src/libstd/sys/windows/stdio.rs

```diff
@@ -58,8 +58,28 @@ fn write(out: &Output, data: &[u8]) -> io::Result<usize> {
         Output::Console(ref c) => c.get().raw(),
         Output::Pipe(ref p) => return p.get().write(data),
     };
-    let utf16 = match str::from_utf8(data).ok() {
-        Some(utf8) => utf8.encode_utf16().collect::<Vec<u16>>(),
+    // As with stdin on windows, stdout often can't handle writes of large
+    // sizes. For an example, see #14940. For this reason, don't try to
+    // write the entire output buffer on windows.
+    //
+    // For some other references, it appears that this problem has been
+    // encountered by others [1] [2]. We choose the number 8K just because
+    // libuv does the same.
+    //
+    // [1]: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1232
+    // [2]: http://www.mail-archive.com/log4net-dev@logging.apache.org/msg00661.html
+    const OUT_MAX: usize = 8192;
+    let (utf16, data_len) = match str::from_utf8(data).ok() {
```

Comment:
Perhaps this could do something like:

```rust
let len = cmp::min(data.len(), OUT_MAX);
let data = match str::from_utf8(&data[..len]) {
    Ok(s) => s,
    Err(ref e) if e.valid_up_to() == 0 => return Err(..),
    Err(e) => str::from_utf8(&s[..e.valid_up_to()]).unwrap(),
};
let data = ...;
```